File: 064_Minimum_Path_Sum.java
// Given a m x n grid filled with non-negative numbers, find a path from top left to bottom right which minimizes the sum of all numbers along its path.

// Note: You can only move either down or right at any point in time.

public class Solution {
    public int minPathSum(int[][] grid) {
        if (grid==null || grid[0] == null){
            return 0;
        }
        int[][] minSum = new int[grid.length][grid[0].length];
        int m = grid.length, n = grid[0].length;
        minSum[0][0] = grid[0][0];
        for (int row = 1; row < m; row ++){
            minSum[row][0] = minSum[row-1][0] + grid[row][0];
        }
        for(int col= 1; col < n; col++){
            minSum[0][col] = minSum[0][col-1] + grid[0][col];
        }
        for (int i = 1; i < m; i++){
            for(int j = 1; j < n ; j++){
                minSum[i][j] = Math.min(minSum[i-1][j], minSum[i][j-1]) + grid[i][j];
            }
        }
        return minSum[m-1][n-1];
    }
}

File: python/064_Minimum_Path_Sum.py
# Given a m x n grid filled with non-negative numbers,
# find a path from top left to bottom right which minimizes
# the sum of all numbers along its path.

# Note: You can only move either down or right at any point in time.

# Subscribe to see which companies asked this question

# Key Point: DP
#            Subproblem: dp[i][j] the minimum path sum to position(i,j)
#            dp[i][j] = min(dp[i-1][j], dp[i][j-1])
# 
# Runtime: 84ms
class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        col, row = len(grid[0]), len(grid)
        dp = [[0 for x in grid[0]] for y in grid]
        for i in xrange(row):
            for j in xrange(col):
                if i==0 and j==0:
                    dp[i][j] = grid[i][j]
                elif i==0:
                    dp[i][j] = dp[i][j-1] + grid[i][j]
                elif j==0:
                    dp[i][j] = dp[i-1][j] + grid[i][j]
                else:
                    dp[i][j] = min(dp[i][j-1], dp[i-1][j]) + grid[i][j]
        return dp[row-1][col-1]

if __name__ == '__main__':
    sol = Solution()
    grid = [[1,2,3], [2,3,4], [3,4,5]]
    print sol.minPathSum(grid)